Mage/src/main/java/mage/game/command/emblems/TyvarKellEmblem.java

```diff
@@ -0,0 +1,103 @@
+package mage.game.command.emblems;
+
+import mage.MageObjectReference;
+import mage.abilities.Ability;
+import mage.abilities.common.SpellCastControllerTriggeredAbility;
+import mage.abilities.effects.ContinuousEffectImpl;
+import mage.abilities.effects.common.DrawCardSourceControllerEffect;
+import mage.abilities.keyword.HasteAbility;
+import mage.cards.Card;
+import mage.constants.*;
+import mage.filter.FilterSpell;
+import mage.game.Game;
+import mage.game.command.Emblem;
+import mage.game.permanent.Permanent;
+import mage.game.stack.Spell;
+
+import java.util.Iterator;
+import java.util.UUID;
+
+/**
+ *
+ * @author weirddan455
+ */
+public final class TyvarKellEmblem extends Emblem {
+
+    private static final FilterSpell filter = new FilterSpell("an Elf spell");
+
+    static {
+        filter.add(SubType.ELF.getPredicate());
+    }
+
+    // −6: You get an emblem with "Whenever you cast an Elf spell, it gains haste until end of turn and you draw two cards."
+    public TyvarKellEmblem() {
+        this.setName("Emblem Tyvar");
+        this.setExpansionSetCodeForImage("KHM");
+        Ability ability = new SpellCastControllerTriggeredAbility(Zone.COMMAND, new TyvarKellEmblemEffect(
+                HasteAbility.getInstance(), Duration.EndOfTurn), filter, false, true
+        );
+        ability.addEffect(new DrawCardSourceControllerEffect(2).setText("and you draw two cards"));
+        this.getAbilities().add(ability);
+    }
+}
+
+class TyvarKellEmblemEffect extends ContinuousEffectImpl {
+
+    protected Ability ability;
+
+    public TyvarKellEmblemEffect(Ability ability, Duration duration) {
+        super(duration, Layer.AbilityAddingRemovingEffects_6, SubLayer.NA, Outcome.AddAbility);
+        this.ability = ability;
+        this.generateGainAbilityDependencies(ability, null);
+        this.staticText = "it gains haste until end of turn";
+    }
+
+    public TyvarKellEmblemEffect(final TyvarKellEmblemEffect effect) {
+        super(effect);
+        this.ability = effect.ability.copy();
+    }
+
+    @Override
+    public TyvarKellEmblemEffect copy() {
+        return new TyvarKellEmblemEffect(this);
+    }
+
+    @Override
+    public void init (Ability source, Game game) {
+        super.init(source, game);
+        if (this.affectedObjectsSet) {
+            Spell spell = game.getStack().getSpell(targetPointer.getFirst(game, source));
+            if (spell != null) {
+                Card card = game.getCard(spell.getSourceId());
+                if (card != null) {
+                    affectedObjectList.add(new MageObjectReference(card, game));
```

Comment:
does this work correctly if the creature gets blinked in the same turn or gets countered and then returned to the battlefield?

Reply:
I just tested with Flicker of Fate.  It keeps the haste after it gets flickered which I think is wrong.

I tested with Remand as well and it shows as having haste after it gets returned to hand.  Not sure if that is correct but I'm thinking no as well.

I think the issue might be that in `apply` I just did `UUID sourceId = it.next().getSourceId();` and I'm not checking for zone change counters.  I'm just not exactly sure how to do that.  It would change zones once going from stack to battlefield right?  So does it start at 0 and I should not apply the effect if it's >1 or how does that work?

Reply:
Actually, I think I found the fix.  I just pushed a new commit that I tested fixes both cases.  It now loses haste if it get blinked or countered.